### PR TITLE
GUA-220 Build, Push and SAM deploy with Github Actions

### PR DESCRIPTION
**Pull request by huwd**
Use [aws-actions/amazon-ecr-login](https://github.com/marketplace/actions/amazon-ecr-login-action-for-github-actions) and a newly created Github Specific OIDC ARN to try and log in, then build our world in docker, push to ECR and Sam Deploy

### CODE REVIEW

File: infrastructure/prototype-app/template.yaml [hunk 1]
```diff
@@ -215,13 +215,39 @@ Resources:
...
 
+  GitHubIdentityProvider:
+    Type: AWS::IAM::OIDCProvider
+    Properties:
+      ClientIdList:
+        - "sts.amazonaws.com"
+      ThumbprintList:
+        - "6938fd4d98bab03faadb97b34396831e3780aea1"
+      Url: "https://token.actions.githubusercontent.com"
+      Tags:
```
Comment (alex9smith) [inline infrastructure/prototype-app/template.yaml:232]: Nice! Is this a standard set of tags we should be using?

File: .github/workflows/build_and_deploy.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+jobs:
+  deploy:
+    name: Deploy POC app to Fargate
+    runs-on: ubuntu-latest
+    timeout-minutes: 60
+    permissions:
+      id-token: write
+      contents: read
+    steps:
+      - name: Set up SAM cli
```
Comment (alex9smith) [inline .github/workflows/build_and_deploy.yaml:17]:
```suggestion
      - name: Checkout
        uses: actions/checkout@v3

      - name: Set up SAM cli
```

Pretty sure you'll also need a checkout action first to get the latest branch contents

**Review by alex9smith — CHANGES_REQUESTED** [review]
This looks like a good start but I don't think it's quite ready.

> I have a sneaking suspicion i've missed a step that gives Github access to docker to run those commands...

`docker` is already installed on the runner machines, so we get that for free 🎉 

I think you've set  `secrets.GH_ACTIONS_ROLE_ARN` as the ARN for the AWS OIDC provider, but that secret needs to be an ARN for a role so you'll need to add a role definintion to the Cloudformation that allows authentication from the new OIDC provider and has permissions to deploy stuff to ECS, push docker images to ECR etc. There's a starting point in the [Github docs](https://github.com/aws-actions/configure-aws-credentials#sample-iam-role-cloudformation-template) but we'll need to also set the permissions on that role.

File: infrastructure/prototype-app/template.yaml (hunk 1 above)
Comment (huwd) [reply infrastructure/prototype-app/template.yaml:232]: I think so.
I wasn't sure what was going on with the ThumbPrint list so I stepped through and verified that was the right hash. Was kinda cool will note below this what I did to prove it.

That URL seems correct.

The rest of the tags are from [here](https://github.com/alphagov/di-devplatform-deploy/blob/main/github-identity/template.yaml). I actually ended up removing two, `Environment` (on the basis we only have one and are not doing build -> integration -> prod for POC), and `System`, which I take to be representing what this is to other parts of DI... again an assumption here that we're less concerned about that 🤔

**Review by alex9smith — APPROVED** [review]
LGTM!  I can't think of a good way of testing this other than merging and hoping, so lets try it🤞🏻

**Pull request merged**